lib/web/app/middleware.go

```diff
@@ -47,14 +50,29 @@ func (h *Handler) withAuth(handler handlerAuthFunc) http.HandlerFunc {
 		// If the caller fails to authenticate, redirect the caller to Teleport.
 		session, err := h.authenticate(r.Context(), r)
 		if err != nil {
-			// If this cluster has the public address of the proxy set, re-direct to
-			// the login page. If not, the best that can be done is to show "403 Forbidden"
-			// because Teleport does not know where to re-direct to.
+			// If this cluster has the public address of the proxy set, redirect to the
+			// login page to redirect to the requested application. If not, the best
+			// that can be done is to show "403 Forbidden" because Teleport does not know
+			// the public address of the Web UI which is needed to launch the application.
 			if h.c.WebPublicAddr != "" {
+				addr, err := utils.ParseAddr(r.Host)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				// While the FQDN will be validated when creating an application session,
+				// make sure we don't send something that doesn't at least look like a
+				// FQDN to the frontend application.
+				if errs := validation.IsDNS1123Subdomain(addr.Host()); len(errs) > 0 {
+					var aerrs []error
+					for _, err := range errs {
+						aerrs = append(aerrs, trace.BadParameter(err))
+					}
+					return trace.NewAggregate(aerrs...)
+				}
 				u := url.URL{
 					Scheme: "https",
 					Host:   h.c.WebPublicAddr,
-					Path:   "/web/login",
+					Path:   fmt.Sprintf("/web/launch/%v", addr.Host()),
```

Comment:
Not sure if I'm missing anything, but this will redirect to the "root" page of the application - so if you e.g. try to go to `grafana.proxy.teleport.io/dashboards/main`, then after login you will get redirected to just `grafana.proxy.teleport.io`? Does this `/launch` handler support redirects to "deep" links?

Reply:
I'm not sure if deep links are supported, that'd be a @alex-kovoy question. That being said, I think it's fine to redirect to the root because this gives you a new "session" similar to an SSH session which starts from a clean slate.

Reply:
I can see deep links being a requested feature, so if it's possible for us to add this then we should at some point.

Reply:
Deep links is a must. For now it redirects back to the app home url.

Reply:
Okay, let's do it in a patch release for 5.0.